proposals/NNNN-explicit-layout-struct.md

```diff
@@ -0,0 +1,187 @@
+
+## Motivation
+
+Some HLSL types have layout that isn't practical to derive from the module's
+DataLayout. This includes all kinds of `cbuffer`, but especially those that use
+`packoffset`, and also applies to structs that use the vulkan `[[vk::offset()]]
+extension and possibly objects with specific alignment specified on subobjects.
+
+We need to be able to represent these types in IR so that we can generate
+correct code in the backends.
+
+## Proposed solution
+
+We should implement a target type that includes a struct type, total size, and
+offsets for each member of the struct. This type can then be used in other
+target types, such as CBuffer or StructuredBuffer definitions, or even in other
+layout types. We need target types for DirectX and for SPIR-V, but the types
+can and should mirror each other.
+
+```
+target("[dx|spirv].Layout", %struct_type, <size>, [offset...])
+```
+
+### Examples
+
+In the examples below we generally use "dx.Layout", since the "spirv.Layout"
+variants would be identical.
+
+While these aren't necessarily needed for types that don't have explicit layout
+rules, some examples of "standard" layout objects represented this way are
+helpful:
+
+```llvm
+; struct simple1 {
+;   float x;
+;   float y;
+; };
+%__hlsl_simple1 = { i32, i32 }
+target("dx.Layout", %__hlsl_simple1, 8, 0, 4)
+
+; struct simple2 {
+;   float3 x;
+;   float y;
+; };
+%__hlsl_simple2 = { <3 x float>, float }
+target("dx.Layout", %__hlsl_simple2, 16, 0, 12)
+
+; struct nested {
+;   simple2 s2;
+;   simple1 s1;
+; };
+%__hlsl_nested = type { target("dx.Layout", %__hlsl_simple2, 16, 0, 12),
+                        target("dx.Layout", %__hlsl_simple1, 8, 0, 4) }
+target("dx.Layout", %__layout_nested, 24, 0, 16)
+```
+
+Objects whose layout differs in cbuffers than in structs:
+
+```llvm
+; struct array_struct {
+;   float x[4];
+;   float y;
+; };
+%__hlsl_array_struct = type { [4 x float], float }
+target("dx.Layout", %__hlsl_array_struct, 20, 0, 16)
+
+; cbuffer array_cbuf1 {
+;   float x[4];
+;   float y;
+; };
+target("dx.Layout", %__hlsl_array_struct, 56, 0, 52)
+
+; cbuffer array_cbuf2 {
+;   array_struct s;
+; };
+target("dx.Layout", %__hlsl_array_struct, 56, 0, 52)
+
+; struct nested2 {
+;   simple1 s1;
+;   simple2 s2;
+; };
+%__hlsl_nested2 = type { target("dx.Layout", %__hlsl_simple1, 8, 0, 4),
+                         target("dx.Layout", %__hlsl_simple2, 16, 0, 12) }
+target("dx.Layout", %__hlsl_nested2, 24, 0, 8)
+
+; cbuffer nested_cbuf {
+;   simple1 s1;
+;   simple2 s2;
+; };
+target("dx.Layout", %__hlsl_nested2, 32, 0, 16)
+```
+
+Simple usage of packoffset:
+
+```llvm
+; cbuffer packoffset1 {
+;   float x : packoffset(c1.x);
+;   float y : packoffset(c2.y);
+; };
+target("dx.Layout", { i32, i32 }, 40, 16, 36)
+```
+
+packoffset that only specifies the first field:
+> note: This emits a warning in DXC. Do we really want to support it?
+
+```llvm
+; cbuffer packoffset1 {
+;   float x : packoffset(c1.x);
+;   float y;
```

Comment:
Isn't the packoffset proposal saying this is not allowed?

Reply:
Same for only specifying the later field.

Reply:
Yes, proposal [0003](https://github.com/llvm/wg-hlsl/blob/main/proposals/0003-packoffset.md), though it disagrees with DXC's behaviour. However, it does match FXC's behaviour and I think it's safe to go forward with disallowing partially defined packoffsets. I'll update the document.